helix-core/src/syntax.rs

```diff
@@ -951,6 +953,86 @@ impl Syntax {
             layers,
             next_event: None,
             last_highlight_range: None,
+            context: (),
+        };
+        result.sort_layers();
+        result
+    }
+
+    /// Iterate over the rainbow-highlighted regions for a given slice of source code.
+    pub fn rainbow_iter<'a>(
+        &'a self,
+        source: RopeSlice<'a>,
+        range: Option<std::ops::Range<usize>>,
+        cancellation_flag: Option<&'a AtomicUsize>,
+        rainbow_length: usize,
+    ) -> impl Iterator<Item = Result<HighlightEvent, Error>> + 'a {
+        let mut layers = self
+            .layers
+            .iter()
+            .filter_map(|(_, layer)| {
+                // TODO: if range doesn't overlap layer range, skip it
+
+                // Reuse a cursor from the pool if available.
+                let mut cursor = PARSER.with(|ts_parser| {
+                    let highlighter = &mut ts_parser.borrow_mut();
+                    highlighter.cursors.pop().unwrap_or_else(QueryCursor::new)
+                });
+
+                // The `captures` iterator borrows the `Tree` and the `QueryCursor`, which
+                // prevents them from being moved. But both of these values are really just
+                // pointers, so it's actually ok to move them.
+                let cursor_ref =
+                    unsafe { mem::transmute::<_, &'static mut QueryCursor>(&mut cursor) };
```

Comment:
Why not instead of using unsafe transmute to extend the lifetime, we wrap this whole thing inside `PARSER.with` so that the lifetime is guaranteed?

Reply:
The borrow checker doesn't like returning the `cursor` and `cursor_ref` or `captures`. I might be misunderstanding you though